pkg/controller/data_upload_controller.go

```diff
@@ -0,0 +1,498 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	snapshotter "github.com/kubernetes-csi/external-snapshotter/client/v4/clientset/versioned/typed/volumesnapshot/v1"
+
+	"github.com/vmware-tanzu/velero/internal/credentials"
+	"github.com/vmware-tanzu/velero/pkg/apis/velero/shared"
+	velerov1api "github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
+	velerov2alpha1api "github.com/vmware-tanzu/velero/pkg/apis/velero/v2alpha1"
+	"github.com/vmware-tanzu/velero/pkg/datapath"
+	"github.com/vmware-tanzu/velero/pkg/exposer"
+	"github.com/vmware-tanzu/velero/pkg/repository"
+	"github.com/vmware-tanzu/velero/pkg/uploader"
+	"github.com/vmware-tanzu/velero/pkg/util/filesystem"
+	"github.com/vmware-tanzu/velero/pkg/util/kube"
+)
+
+const dataMoverType string = "velero"
+const dataUploadDownloadRequestor string = "snapshot-data-upload-download"
+
+// DataUploadReconciler reconciles a DataUpload object
+type DataUploadReconciler struct {
+	client              client.Client
+	kubeClient          kubernetes.Interface
+	csiSnapshotClient   snapshotter.SnapshotV1Interface
+	repoEnsurer         *repository.Ensurer
+	clock               clocks.WithTickerAndDelayedExecution
+	credentialGetter    *credentials.CredentialGetter
+	nodeName            string
+	fileSystem          filesystem.Interface
+	logger              logrus.FieldLogger
+	snapshotExposerList map[velerov2alpha1api.SnapshotType]exposer.SnapshotExposer
+	dataPathMgr         *datapath.Manager
+}
+
+func NewDataUploadReconciler(client client.Client, kubeClient kubernetes.Interface,
+	csiSnapshotClient snapshotter.SnapshotV1Interface, repoEnsurer *repository.Ensurer, clock clocks.WithTickerAndDelayedExecution,
+	cred *credentials.CredentialGetter, nodeName string, fs filesystem.Interface, log logrus.FieldLogger) *DataUploadReconciler {
+	return &DataUploadReconciler{
+		client:              client,
+		kubeClient:          kubeClient,
+		csiSnapshotClient:   csiSnapshotClient,
+		clock:               clock,
+		credentialGetter:    cred,
+		nodeName:            nodeName,
+		fileSystem:          fs,
+		logger:              log,
+		repoEnsurer:         repoEnsurer,
+		snapshotExposerList: map[velerov2alpha1api.SnapshotType]exposer.SnapshotExposer{velerov2alpha1api.SnapshotTypeCSI: exposer.NewCSISnapshotExposer(kubeClient, csiSnapshotClient, log)},
+		dataPathMgr:         datapath.NewManager(1),
+	}
+}
+
+// +kubebuilder:rbac:groups=velero.io,resources=dataupload,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=velero.io,resources=dataupload/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups="",resources=pods,verbs=get
+// +kubebuilder:rbac:groups="",resources=persistentvolumes,verbs=get
+// +kubebuilder:rbac:groups="",resources=persistentvolumerclaims,verbs=get
```

Comment:
Need a blank line between the marker comment and the Reconcile function.

Reply:
Done